README.md
# Jekyll::Gist

Liquid tag for displaying GitHub Gists in Jekyll sites: `{% gist %}`.

[![Build Status](https://travis-ci.org/jekyll/jekyll-gist.svg?branch=master)](https://travis-ci.org/jekyll/jekyll-gist)

## Installation

Add this line to your application's Gemfile:

    $ gem 'jekyll-gist'

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install jekyll-gist

Then add the following to your site's `_config.yml`:

```
plugins:
  - jekyll-gist
```

💡 If you are using a Jekyll version less than 3.5.0, use the `gems` key instead of `plugins`.

## Usage

Use the tag as follows in your Jekyll pages, posts and collections:

```liquid
{% gist c08ee0f2726fd0e3909d %}
```

This will create the associated script tag:

```html
<script src="https://gist.github.com/parkr/c08ee0f2726fd0e3909d.js"> </script>
```

You may optionally specify a `filename` after the `gist_id`:

```liquid
{% gist c08ee0f2726fd0e3909d test.md %}
```

This will produce the correct URL to show just the specified file in your post rather than the entire Gist.

**Pro-tip**: If you provide a personal access token with Gist scope, as the environmental variable `JEKYLL_GITHUB_TOKEN`, Jekyll Gist will use the Gist API to speed up site generation.

## Disabling `noscript` support

By default, Jekyll Gist will make an HTTP call per Gist to retrieve the raw content of the Gist. This information is used to propagate `noscript` tags for search engines and browsers without JavaScript support. If you'd like to disable this feature, for example, to speed up builds locally, add the following to your site's `_config.yml`:

```yml
gist:
  noscript: false
```

## Contributing

1. Fork it ( https://github.com/jekyll/jekyll-gist/fork )
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create a new Pull Request


History.markdown
## HEAD

### Development Fixes

  * Test against Ruby 2.5 (#57)
  * Rely on rubocop-jekyll (#62)
  * chore(deps): rubocop-jekyll 0.3 (#64)
  * Refactor GistTag (#65)

### Major Enhancements

  * chore(deps): drop support for Liquid &lt; 4.0 (#66)

### Bug Fixes

  * Re-introduce Ruby 2.3 support and test Jekyll 3.7+ (#72)
  * Use Liquid::Tag#raw to clarify error message (#73)

## 1.5.0 / 2017-12-03

### Documentation

  * replace &#39;plugins&#39; key in config with &#39;gems&#39; (#46)
  * Docs: Remove username in gist (#54)

### Development Fixes

  * Remove testing for Jekyll 2.x
  * Requires Ruby > 2.1
  * Add release script
  * Inherit Jekyll&#39;s rubocop config for consistency (#48)
  * define path with __dir__ (#47)

## 1.4.1 / 2017-06-21

  * Don't ask .empty? until it's a String. (#38)
  * rename Liquid 4 `has_key?` to `key?` to add compatibility for liquid 4 (#41)
  * Test against Ruby 2.1 to 2.4 (#45)

## 1.4.0 / 2015-12-01

  * Allow `noscript` fallback to be disabled (#29)
  * Use Octokit to fetch Gist content when passed `JEKYLL_GITHUB_TOKEN` in env(#28)

## 1.3.5 / 2015-10-23

  * Fix encoding error for `noscript` code (#23)
  * Test against Jekyll 3, 2, and the github-pages gem (#19)

## 1.3.4 / 2015-08-28

  * Catch `TimeoutError` to further support 1.9.3 (#16)

## 1.3.3 / 2015-08-20

  * Fix gemspec to allow Ruby 1.9.3 (relates to #14)

## 1.3.2 / 2016-08-19

  * Re-add support for Ruby 1.9.3. Fixes #11 for 1.9.3 (#14)
  * Replaced `OpenURI` with `Net::HTTP` and introduced timeout of 3 seconds (#11)

## 1.3.1 / 2015-08-16

  * Replaced `OpenURI` with `Net::HTTP` and introduced timeout of 3 seconds (#11)

## 1.3.0 / 2015-08-05

  * Added an `noscript` fallback for browsers without JavaScript enabled. (#7)

## 1.2.1 / 2015-03-22

  * Use `has_key?` (#6)

## 1.2.0 / 2015-03-21

### Minor Enhancements

  * Allow variables as parameters (#4)

### Development Fixes

  * Fix RSpec deprecation warning (#5)

## 1.1.0 / 2014-06-18

### Minor Enhancements

  * Update regex to allow for new sha-ish ids in Gist. (#1)

## 1.0.0 / 2014-06-01

  * Birthday!


spec/spec_helper.rb
TEST_DIR = __dir__
TMP_DIR  = File.expand_path("../tmp", TEST_DIR)

require "webmock/rspec"
require "cgi"
require "jekyll"
require File.expand_path("../lib/jekyll-gist.rb", TEST_DIR)

Jekyll.logger.log_level = :error

RSpec.configure do |config|
  config.run_all_when_everything_filtered = true
  config.filter_run :focus
  config.order = "random"

  def tmp_dir(*files)
    File.join(TMP_DIR, *files)
  end

  def source_dir(*files)
    tmp_dir("source", *files)
  end

  def dest_dir(*files)
    tmp_dir("dest", *files)
  end

  def doc_with_content(_content, opts = {})
    my_site = site(opts)
    Jekyll::Document.new(source_dir("_test/doc.md"), { :site => my_site, :collection => collection(my_site) })
  end

  def collection(site, label = "test")
    Jekyll::Collection.new(site, label)
  end

  def site(opts = {})
    conf = Jekyll::Utils.deep_merge_hashes(Jekyll::Configuration::DEFAULTS, opts.merge({
      "source"      => source_dir,
      "destination" => dest_dir,
    }))
    Jekyll::Site.new(conf)
  end

  def fixture(name)
    path = File.expand_path "fixtures/#{name}.json", __dir__
    File.open(path).read
  end
end


spec/gist_tag_spec.rb
require "spec_helper"

describe(Jekyll::Gist::GistTag) do
  let(:http_output) { "<test>true</test>" }
  let(:doc) { doc_with_content(content) }
  let(:content) { "{% gist #{gist} %}" }
  let(:output) do
    doc.content = content
    doc.output  = Jekyll::Renderer.new(doc.site, doc).run
  end

  before(:each) { ENV["JEKYLL_GITHUB_TOKEN"] = nil }

  context "valid gist" do
    context "with user prefix" do
      before { stub_request(:get, "https://gist.githubusercontent.com/#{gist}/raw").to_return(:body => http_output) }
      let(:gist) { "mattr-/24081a1d93d2898ecf0f" }

      it "produces the correct script tag" do
        expect(output).to match(%r!<script src="https:\/\/gist.github.com\/#{gist}.js">\s<\/script>!)
      end
      it "produces the correct noscript tag" do
        expect(output).to match(%r!<noscript><pre>&lt;test&gt;true&lt;\/test&gt;<\/pre><\/noscript>\n!)
      end
    end

    context "without user prefix" do
      before { stub_request(:get, "https://gist.githubusercontent.com/#{gist}/raw").to_return(:body => http_output) }
      let(:gist) { "28949e1d5ee2273f9fd3" }

      it "produces the correct script tag" do
        expect(output).to match(%r!<script src="https:\/\/gist.github.com\/#{gist}.js">\s<\/script>!)
      end
      it "produces the correct noscript tag" do
        expect(output).to match(%r!<noscript><pre>&lt;test&gt;true&lt;\/test&gt;<\/pre><\/noscript>\n!)
      end
    end

    context "classic Gist id style" do
      before { stub_request(:get, "https://gist.githubusercontent.com/#{gist}/raw").to_return(:body => http_output) }
      let(:gist) { "1234321" }

      it "produces the correct script tag" do
        expect(output).to match(%r!<script src="https:\/\/gist.github.com\/#{gist}.js">\s<\/script>!)
      end
      it "produces the correct noscript tag" do
        expect(output).to match(%r!<noscript><pre>&lt;test&gt;true&lt;\/test&gt;<\/pre><\/noscript>\n!)
      end
    end

    context "with file specified" do
      before { stub_request(:get, "https://gist.githubusercontent.com/#{gist}/raw/#{filename}").to_return(:body => http_output) }
      let(:gist)     { "mattr-/24081a1d93d2898ecf0f" }
      let(:filename) { "myfile.ext" }
      let(:content)  { "{% gist #{gist} #{filename} %}" }

      it "produces the correct script tag" do
        expect(output).to match(%r!<script src="https:\/\/gist.github.com\/#{gist}.js\?file=#{filename}">\s<\/script>!)
      end
      it "produces the correct noscript tag" do
        expect(output).to match(%r!<noscript><pre>&lt;test&gt;true&lt;\/test&gt;<\/pre><\/noscript>\n!)
      end
    end

    context "with variable gist id" do
      before { stub_request(:get, "https://gist.githubusercontent.com/#{gist_id}/raw").to_return(:body => http_output) }
      let(:gist_id)  { "1342013" }
      let(:gist)     { "page.gist_id" }
      let(:output) do
        doc.data["gist_id"] = gist_id
        doc.content = content
        doc.output  = Jekyll::Renderer.new(doc.site, doc).run
      end

      it "produces the correct script tag" do
        expect(output).to match(%r!<script src="https:\/\/gist.github.com\/#{doc.data['gist_id']}.js">\s<\/script>!)
      end
      it "produces the correct noscript tag" do
        expect(output).to match(%r!<noscript><pre>&lt;test&gt;true&lt;\/test&gt;<\/pre><\/noscript>\n!)
      end
    end

    context "with variable gist id and filename" do
      before { stub_request(:get, "https://gist.githubusercontent.com/#{gist_id}/raw/#{gist_filename}").to_return(:body => http_output) }
      let(:gist_id)       { "1342013" }
      let(:gist_filename) { "atom.xml" }
      let(:gist)          { "page.gist_id" }
      let(:filename)      { "page.gist_filename" }
      let(:content)       { "{% gist #{gist} #{filename} %}" }
      let(:output) do
        doc.data["gist_id"] = "1342013"
        doc.data["gist_filename"] = "atom.xml"
        doc.content = content
        doc.output  = Jekyll::Renderer.new(doc.site, doc).run
      end

      it "produces the correct script tag" do
        expect(output).to match(%r!<script src="https:\/\/gist.github.com\/#{doc.data['gist_id']}.js\?file=#{doc.data['gist_filename']}">\s<\/script>!)
      end

      it "produces the correct noscript tag" do
        expect(output).to match(%r!<noscript><pre>&lt;test&gt;true&lt;\/test&gt;<\/pre><\/noscript>\n!)
      end
    end

    context "with valid gist id and invalid filename" do
      before { stub_request(:get, "https://gist.githubusercontent.com/#{gist_id}/raw/#{gist_filename}").to_return(:status => 404) }
      let(:gist_id) { "mattr-/24081a1d93d2898ecf0f" }
      let(:gist_filename) { "myfile.ext" }
      let(:content) { "{% gist #{gist_id} #{gist_filename} %}" }

      it "produces the correct script tag" do
        expect(output).to match(%r!<script src="https:\/\/gist.github.com\/#{gist_id}.js\?file=#{gist_filename}">\s<\/script>!)
      end

      it "does not produce the noscript tag" do
        expect(output).to_not match(%r!<noscript><pre>&lt;test&gt;true&lt;\/test&gt;<\/pre><\/noscript>\n!)
      end
    end

    context "with token" do
      before { ENV["JEKYLL_GITHUB_TOKEN"] = "1234" }
      before do
        stub_request(:get, "https://api.github.com/gists/1342013")
          .to_return(:status => 200, :body => fixture("single-file"), :headers => { "Content-Type" => "application/json" })
      end
      let(:gist_id)  { "1342013" }
      let(:gist)     { "page.gist_id" }
      let(:output) do
        doc.data["gist_id"] = gist_id
        doc.content = content
        doc.output  = Jekyll::Renderer.new(doc.site, doc).run
      end

      it "produces the noscript tag" do
        expect(output).to match(%r!<noscript><pre>contents of gist<\/pre><\/noscript>!)
      end

      context "with a filename" do
        before do
          stub_request(:get, "https://api.github.com/gists/1342013")
            .to_return(:status => 200, :body => fixture("multiple-files"), :headers => { "Content-Type" => "application/json" })
        end
        let(:content) { "{% gist 1342013 hello-world.rb %}" }

        it "produces the noscript tag" do
          expect(output).to match(%r!<noscript><pre>puts &#39;hello world&#39;<\/pre><\/noscript>!)
        end
      end
    end

    context "with noscript disabled" do
      let(:doc) { doc_with_content(content, { "gist" => { "noscript" => false } }) }
      let(:output) do
        doc.content = content
        doc.output  = Jekyll::Renderer.new(doc.site, doc).run
      end
      let(:gist) { "mattr-/24081a1d93d2898ecf0f" }

      it "does not produce the noscript tag" do
        expect(output).to_not match(%r!<noscript>!)
      end
    end
  end

  context "invalid gist" do
    context "no gist id present" do
      let(:gist) { "" }

      it "raises an error" do
        expect(-> { output }).to raise_error(ArgumentError)
      end
    end
  end
end


lib/jekyll-gist.rb
require "jekyll/gist/version"

module Jekyll
  class Gist < Liquid::Tag

    def render(context)
      if tag_contents = determine_arguments(@markup.strip)
        gist_id, filename = tag_contents[0], tag_contents[1]
        gist_script_tag(gist_id, filename)
      else
        raise ArgumentError.new <<-eos
Syntax error in tag 'gist' while parsing the following markup:

  #{@markup}

Valid syntax:
  for all gists: {% gist user/1234567 %}
eos
      end
    end

    private

    def determine_arguments(input)
      matched = if input.include?("/")
        input.match(/\A([a-zA-Z0-9\/\-_]+) ?(\S*)\Z/)
      else
        input.match(/\A(\d+) ?(\S*)\Z/)
      end
      [matched[1].strip, matched[2].strip] if matched && matched.length >= 3
    end

    def gist_script_tag(gist_id, filename = nil)
      if filename.empty?
        "<script src=\"https://gist.github.com/#{gist_id}.js\"> </script>"
      else
        "<script src=\"https://gist.github.com/#{gist_id}.js?file=#{filename}\"> </script>"
      end
    end
  end
end

Liquid::Template.register_tag('gist', Jekyll::GistTag)


lib/jekyll-gist/gist_tag.rb
# frozen_string_literal: true

require "cgi"
require "net/http"
require "octokit"

Net::OpenTimeout = Class.new(RuntimeError) unless Net.const_defined?(:OpenTimeout)
Net::ReadTimeout = Class.new(RuntimeError) unless Net.const_defined?(:ReadTimeout)

module Jekyll
  module Gist
    class GistTag < Liquid::Tag
      def self.client
        @client ||= Octokit::Client.new :access_token => ENV["JEKYLL_GITHUB_TOKEN"]
      end

      def render(context)
        @encoding = context.registers[:site].config["encoding"] || "utf-8"
        @settings = context.registers[:site].config["gist"]
        if (tag_contents = determine_arguments(@markup.strip))
          gist_id  = tag_contents[0]
          filename = tag_contents[1]
          gist_id  = context[gist_id]  if context.key?(gist_id)
          filename = context[filename] if context.key?(filename)

          noscript_tag = gist_noscript_tag(gist_id, filename)
          script_tag   = gist_script_tag(gist_id, filename)

          "#{noscript_tag}#{script_tag}"
        else
          raise ArgumentError, <<~ERROR
            Syntax error in tag 'gist' while parsing the following markup:

              '{% #{raw.strip} %}'

            Valid syntax:
              {% gist user/1234567 %}
              {% gist user/1234567 foo.js %}
              {% gist 28949e1d5ee2273f9fd3 %}
              {% gist 28949e1d5ee2273f9fd3 best.md %}

          ERROR
        end
      end

      private

      def determine_arguments(input)
        matched = input.match(%r!\A([\S]+|.*(?=\/).+)\s?(\S*)\Z!)
        [matched[1].strip, matched[2].strip] if matched && matched.length >= 3
      end

      def gist_script_tag(gist_id, filename = nil)
        url = "https://gist.github.com/#{gist_id}.js"
        url = "#{url}?file=#{filename}" unless filename.to_s.empty?

        "<script src=\"#{url}\"> </script>"
      end

      def gist_noscript_tag(gist_id, filename = nil)
        return if @settings && @settings["noscript"] == false

        code = fetch_raw_code(gist_id, filename)
        if code
          code = code.force_encoding(@encoding)
          code = CGI.escapeHTML(code).gsub("'", "&#39;")

          "<noscript><pre>#{code}</pre></noscript>"
        else
          Jekyll.logger.warn "Warning:", "The <noscript> tag for your gist #{gist_id}"
          Jekyll.logger.warn "", "could not be generated. This will affect users who do"
          Jekyll.logger.warn "", "not have JavaScript enabled in their browsers."
        end
      end

      def fetch_raw_code(gist_id, filename = nil)
        return code_from_api(gist_id, filename) if ENV["JEKYLL_GITHUB_TOKEN"]

        url = "https://gist.githubusercontent.com/#{gist_id}/raw"
        url = "#{url}/#{filename}" unless filename.to_s.empty?
        uri = URI(url)

        Net::HTTP.start(uri.host, uri.port,
                        :use_ssl => uri.scheme == "https",
                        :read_timeout => 3, :open_timeout => 3) do |http|
          request = Net::HTTP::Get.new uri.to_s
          response = http.request(request)
          response.body
        end
      rescue SocketError, Net::HTTPError, Net::OpenTimeout, Net::ReadTimeout, TimeoutError
        nil
      end

      def code_from_api(gist_id, filename = nil)
        gist = GistTag.client.gist gist_id
        file = if filename.to_s.empty?
                 # No file specified, return the value of the first key/value pair
                 gist.files.first[1]
               else
                 # .files is a hash of :"filename.extension" => data pairs
                 # Rather than using to_sym on arbitrary user input,
                 # Find our file by calling to_s on the keys
                 match = gist.files.find { |name, _data| name.to_s == filename }
                 match[1] if match
               end

        file[:content] if file
      end
    end
  end
end

Liquid::Template.register_tag("gist", Jekyll::Gist::GistTag)
